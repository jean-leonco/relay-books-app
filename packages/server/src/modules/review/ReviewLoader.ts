import DataLoader from 'dataloader';
import { connectionFromMongoAggregate, mongooseLoader } from '@entria/graphql-mongoose-loader';
import { ConnectionArguments } from 'graphql-relay';
import { Types } from 'mongoose';
import { buildMongoConditionsFromFilters } from '@entria/graphql-mongo-helpers';

import { GraphQLContext, DataLoaderKey } from '../../types';

import { NullConnection } from '../../graphql/connection/NullConnection';
import { buildAggregatePipeline } from '../../core/graphql/graphqlFilters/graphqlFilters';

import ReviewModel, { IReview } from './ReviewModel';
import { reviewFilterMapping, ReviewsArgFilters } from './filters/ReviewFiltersInputType';

export default class Review {
  public registeredType = 'Review';

  id: string;
  _id: Types.ObjectId;
  bookId: Types.ObjectId;
  userId: Types.ObjectId;
  rating: number;
  description?: string;
  isActive: boolean;
  removedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;

  constructor(data: IReview) {
    this.id = data.id || data._id;
    this._id = data._id;
    this.bookId = data.bookId;
    this.userId = data.userId;
    this.rating = data.rating;
    this.description = data.description;
    this.isActive = data.isActive;
    this.removedAt = data.removedAt;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }
}

export const getLoader = () => new DataLoader((ids) => mongooseLoader(ReviewModel, ids));

const viewerCanSee = (context: GraphQLContext, data: IReview) => {
  if (!context.user) {
    return false;
  }

  if (!data.isActive || data.removedAt) {
    return false;
  }

  return true;
};

export const load = async (context: GraphQLContext, id: DataLoaderKey) => {
  if (!id) return null;

  try {
    const data = await context.dataloaders.ReviewLoader.load(id);

    if (!data) return null;

    return viewerCanSee(context, data) ? new Review(data) : null;
  } catch (err) {
    return null;
  }
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) =>
  dataloaders.ReviewLoader.clear(id.toString());

export const primeCache = ({ dataloaders }: GraphQLContext, id: string, data: IReview) =>
  dataloaders.ReviewLoader?.prime(id.toString(), data);

export const clearAndPrimeCache = (context: GraphQLContext, id: string, data: IReview) =>
  clearCache(context, id) && primeCache(context, id, data);

interface LoadReviewsArgs extends ConnectionArguments {
  filters?: ReviewsArgFilters;
}
export const loadReviews = async (context: GraphQLContext, args: LoadReviewsArgs) => {
  const { user } = context;

  if (!user) {
    return NullConnection;
  }

  const { filters = {} } = args;

  const defaultFilters = { orderBy: [{ field: 'createdAt', direction: -1 }] };
  const defaultConditions = { isActive: true, removedAt: null };

  const builtMongoConditions = buildMongoConditionsFromFilters(
    context,
    { ...defaultFilters, ...filters },
    reviewFilterMapping,
  );

  const aggregatePipeline = buildAggregatePipeline({ defaultConditions, builtMongoConditions });

  const aggregate = ReviewModel.aggregate(aggregatePipeline);

  return connectionFromMongoAggregate({
    aggregate,
    context,
    args,
    loader: load,
  });
};

export const loadBookAverageRating = async (context: GraphQLContext, bookId: Types.ObjectId) => {
  const { user } = context;

  if (!user) {
    return 0;
  }

  const aggregate = await ReviewModel.aggregate<{ average: number }>([
    { $match: { bookId } },
    {
      $group: {
        _id: null,
        totalRating: { $sum: '$rating' },
        count: { $sum: 1 },
      },
    },
    {
      $project: {
        average: { $divide: ['$totalRating', '$count'] },
      },
    },
  ]);

  // in this case, there is no review for the provided id
  if (aggregate.length === 0) {
    return 0;
  }

  return aggregate[0].average;
};
