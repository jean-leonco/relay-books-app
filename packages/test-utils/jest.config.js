const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTestFramework.js'],
  globalSetup: '<rootDir>/test/setup.js',
  globalTeardown: '<rootDir>/test/teardown.js',
  transform: {
    // '^.+\\.(js|jsx|ts|tsx)?$': require.resolve('babel-jest'),
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/test/babel-transformer',
  },
};
