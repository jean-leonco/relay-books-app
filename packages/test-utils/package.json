{
  "name": "@booksapp/test-utils",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@booksapp/babel": "^0.0.1",
    "@booksapp/types": "^0.0.1",
    "@koa/router": "^9.3.1",
    "babel-jest": "^26.1.0",
    "graphql-relay": "0.6.0",
    "koa": "^2.13.0",
    "koa-bodyparser": "^4.3.0",
    "mongoose": "^5.10.7",
    "supertest": "^5.0.0"
  }
}
