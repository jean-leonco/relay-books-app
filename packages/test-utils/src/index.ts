export { defaultFrozenKeys, sanitizeValue, sanitizeTestObject } from './objectNormalization';
export { httpRequestGraphql } from './httpRequestGraphql';
