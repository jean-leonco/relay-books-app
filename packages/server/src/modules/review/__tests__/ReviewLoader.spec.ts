import { sanitizeTestObject } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  getContext,
  createReview,
  createUser,
  createBook,
} from '../../../../test/helpers';

import * as ReviewLoader from '../ReviewLoader';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('ReviewLoader', () => {
  it('should be able to load an review', async () => {
    const user = await createUser();
    const review = await createReview();

    const context = await getContext({ user });
    const reviewObj = await ReviewLoader.load(context, review._id);

    expect(reviewObj).not.toBe(null);
    expect(reviewObj!.bookId).toEqual(review.bookId);
    expect(reviewObj!.rating).toEqual(review.rating);
    expect(reviewObj!.userId).toMatchObject(review.userId);
  });

  it('should not be able to load an review without context.user', async () => {
    const review = await createReview();

    const context = await getContext({});
    const reviewObj = await ReviewLoader.load(context, review._id);

    expect(reviewObj).toBe(null);
  });

  it('should be able to load a list of reviews', async () => {
    const user = await createUser();

    for (let i = 0; i < 11; i++) {
      await createReview();
    }

    const context = await getContext({ user });
    const reviewObj = await ReviewLoader.loadReviews(context, {});

    expect(reviewObj).not.toBe(null);
    expect(reviewObj.count).toBe(11);
    expect(sanitizeTestObject(reviewObj)).toMatchSnapshot();
  });

  it('should not be able to load a list of reviews without context.user', async () => {
    for (let i = 0; i < 11; i++) {
      await createReview();
    }

    const context = await getContext();
    const reviewObj = await ReviewLoader.loadReviews(context, {});

    expect(reviewObj).not.toBe(null);
    expect(reviewObj.count).toBe(0);
  });

  it('should be able to load a book rating', async () => {
    const user = await createUser();
    const book = await createBook();

    let totalRating = 0;
    const totalReviews = 5;

    for (let i = 0; i < totalReviews; i++) {
      const rating = i % 2 === 0 ? 5 : 2;
      await createReview({ rating });
      totalRating += rating;
    }

    const context = await getContext({ user });
    const bookAverage = await ReviewLoader.loadBookAverageRating(context, book._id);

    expect(bookAverage).not.toBe(null);
    expect(bookAverage).toBe(totalRating / totalReviews);
  });

  it('should be able to load a book rating even if there is no review', async () => {
    const user = await createUser();
    const book = await createBook();

    const context = await getContext({ user });
    const bookAverage = await ReviewLoader.loadBookAverageRating(context, book._id);

    expect(bookAverage).not.toBe(null);
    expect(bookAverage).toBe(0);
  });

  it('should not be able to load a book rating without an user', async () => {
    const book = await createBook();

    const context = await getContext();
    const bookAverage = await ReviewLoader.loadBookAverageRating(context, book._id);

    expect(bookAverage).not.toBe(null);
    expect(bookAverage).toBe(0);
  });
});
