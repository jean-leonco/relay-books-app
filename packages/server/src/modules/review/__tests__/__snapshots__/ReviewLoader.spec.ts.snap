// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewLoader should be able to load a list of reviews 1`] = `
Object {
  "count": 11,
  "edges": Array [
    Object {
      "cursor": "bW9uZ286MA==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286MQ==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286Mg==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286Mw==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286NA==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286NQ==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286Ng==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286Nw==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286OA==",
      "node": Object {},
    },
    Object {
      "cursor": "bW9uZ286OQ==",
      "node": Object {},
    },
  ],
  "endCursorOffset": 10,
  "pageInfo": Object {
    "endCursor": "bW9uZ286OQ==",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "bW9uZ286MA==",
  },
  "startCursorOffset": 0,
}
`;
